File: seed/devprod/starlark/inbazel/go/inbazel/inbazel.go
package inbazel

import (
	"io"
	"os"
	"path/filepath"
	"strings"

	"github.com/ndscm/theseed/seed/infra/error/go/seederr"
)

// CheckSandbox checks if we are running in a bazel linux sandbox
// by looking for a read-only root mount.
func CheckSandbox() (bool, error) {
	mounts, err := os.ReadFile("/proc/self/mounts")
	if err != nil {
		return false, seederr.Wrap(err)
	}
	for _, line := range strings.Split(string(mounts), "\n") {
		fields := strings.Fields(line)
		if len(fields) >= 4 && fields[1] == "/" {
			for _, mode := range strings.Split(fields[3], ",") {
				if mode == "ro" {
					return true, nil
				}
			}
			return false, nil
		}
	}
	return false, nil
}

// Unlink resolves a bazel-wrapped symlink. When sandbox is true, it
// reads the symlink target and joins it with the parent directory.
func Unlink(src string) (string, error) {
	srcStat, err := os.Lstat(src)
	if err != nil {
		return "", seederr.Wrap(err)
	}
	if srcStat.Mode()&os.ModeSymlink == 0 {
		return "", seederr.WrapErrorf("expected symlink, got regular file (not in bazel sandbox?): %s", src)
	}
	rawPath, err := os.Readlink(src)
	if err != nil {
		return "", seederr.Wrap(err)
	}
	srcPath := filepath.Join(filepath.Dir(src), rawPath)
	return srcPath, nil
}

func CopyFile(src string, dst string) error {
	srcReader, err := os.Open(src)
	if err != nil {
		return seederr.Wrap(err)
	}
	defer srcReader.Close()
	srcStat, err := srcReader.Stat()
	if err != nil {
		return seederr.Wrap(err)
	}
	err = os.Remove(dst)
	if err != nil && !os.IsNotExist(err) {
		return seederr.Wrap(err)
	}
	dstWriter, err := os.OpenFile(dst, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, srcStat.Mode())
	if err != nil {
		return seederr.Wrap(err)
	}
	defer dstWriter.Close()
	_, err = io.Copy(dstWriter, srcReader)
	if err != nil {
		return seederr.Wrap(err)
	}
	return nil
}

// CopyTree recursively copies a directory tree from src to dst,
// resolving bazel symlinks when sandbox is true.
func CopyTree(src string, dst string, sandbox bool) error {
	entries, err := os.ReadDir(src)
	if err != nil {
		return seederr.Wrap(err)
	}
	for _, entryStat := range entries {
		s := filepath.Join(src, entryStat.Name())
		d := filepath.Join(dst, entryStat.Name())
		if entryStat.IsDir() {
			err = os.MkdirAll(d, 0755)
			if err != nil {
				return seederr.Wrap(err)
			}
			err = CopyTree(s, d, sandbox)
			if err != nil {
				return seederr.Wrap(err)
			}
			continue
		}
		resolved := s
		if sandbox {
			resolved, err = Unlink(s)
			if err != nil {
				return seederr.Wrap(err)
			}
		}
		err = CopyFile(resolved, d)
		if err != nil {
			return seederr.Wrap(err)
		}
	}
	return nil
}

// CopySrcs copies a list of source files/directories into dstBaseDir.
// stripComponents controls how much of each src path prefix is removed:
//   - >= 0: strip the first N path components (like tar --strip-components).
//   - < 0: flatten — discard the entire src path and place files directly
//     under dstBaseDir using only their basename.
func CopySrcs(srcs []string, dstBaseDir string, stripComponents int, sandbox bool) error {
	for _, src := range srcs {
		dst := src
		if stripComponents < 0 {
			dst = ""
		} else {
			parts := strings.SplitN(dst, string(filepath.Separator), stripComponents+1)
			dst = parts[len(parts)-1]
		}
		realDst := filepath.Join(dstBaseDir, dst)
		err := os.MkdirAll(filepath.Dir(realDst), 0755)
		if err != nil {
			return seederr.Wrap(err)
		}
		srcStat, err := os.Stat(src)
		if err != nil {
			return seederr.Wrap(err)
		}
		if srcStat.IsDir() {
			err = CopyTree(src, realDst, sandbox)
			if err != nil {
				return seederr.Wrap(err)
			}
			continue
		}
		resolved := src
		if sandbox {
			resolved, err = Unlink(src)
			if err != nil {
				return seederr.Wrap(err)
			}
		}
		err = CopyFile(resolved, realDst)
		if err != nil {
			return seederr.Wrap(err)
		}
	}
	return nil
}

File: seed/devprod/starlark/inbazel/python/inbazel.py
import os
import shutil


def check_sandbox() -> bool:
    """Check if we are running in a bazel linux sandbox by looking for a read-only root mount."""
    with open("/proc/self/mounts", "r") as f:
        mounts = f.read()
    for line in mounts.splitlines():
        mount = line.split()
        if len(mount) >= 4 and mount[1] == "/":
            modes = mount[3].split(",")
            if "ro" in modes:
                return True
            return False
    return False


def unlink(src: str) -> str:
    """Resolve a bazel-wrapped symlink. Raises if src is not a symlink."""
    if not os.path.islink(src):
        raise RuntimeError(
            f"expected symlink, got regular file (not in bazel sandbox?): {src}"
        )
    return os.path.join(os.path.dirname(src), os.readlink(src))


def copy_tree(src: str, dst: str, sandbox: bool = True):
    """Recursively copy a directory tree, resolving bazel symlinks when sandbox is True."""
    for basename in os.listdir(src):
        s = os.path.join(src, basename)
        d = os.path.join(dst, basename)
        if os.path.isdir(s):
            os.makedirs(d, exist_ok=True)
            copy_tree(s, d, sandbox=sandbox)
        else:
            resolved = unlink(s) if sandbox else s
            shutil.copy(resolved, d, follow_symlinks=False)


def copy_srcs(
    srcs: list[str],
    target_folder: str,
    strip_components: int = 0,
    sandbox: bool = True,
):
    """Copy source files/directories into target_folder.

    strip_components controls how much of each src path prefix is removed:
      >= 0: strip the first N path components (like tar --strip-components).
      < 0: flatten — discard the entire src path and place files directly
           under target_folder using only their basename.
    """
    for src in srcs:
        dst = src if strip_components >= 0 else ""
        dst = dst.split(os.sep, strip_components)[-1]
        real_dst = os.path.join(target_folder, dst)
        os.makedirs(os.path.dirname(real_dst), exist_ok=True)
        if os.path.isdir(src):
            copy_tree(src, real_dst, sandbox=sandbox)
            continue
        resolved = unlink(src) if sandbox else src
        shutil.copy(resolved, real_dst, follow_symlinks=False)
